drivers/timer/ite_it8xxx2_timer.c

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (c) 2020 ITE Corporation. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Jyunlin Chen <jyunlin.chen@ite.com.tw>
+ */
+
+#include <zephyr.h>
+#include <sys/util.h>
+#include <drivers/timer/system_timer.h>
+#include <soc.h>
+#include <sys/printk.h>
+/**
+ * Macro Define
+ */
+#define EXT_CTL_B           (EC_Register_Base_Address + 0x1F10)
+#define EXT_PSC_B           (EC_Register_Base_Address + 0x1F11)
+#define EXT_LLR_B           (EC_Register_Base_Address + 0x1F14)
+#define EXT_LHR_B           (EC_Register_Base_Address + 0x1F15)
+#define EXT_LH2R_B          (EC_Register_Base_Address + 0x1F16)
+#define EXT_LH3R_B          (EC_Register_Base_Address + 0x1F17)
+#define EXT_CNTO_B          (EC_Register_Base_Address + 0x1F48)
+
+#define CTIMER_HW_TIMER_INDEX           EXT_TIMER_3
+#define ETIMER_HW_TIMER_INDEX           EXT_TIMER_5
+#define RTIMER_HW_TIMER_INDEX           EXT_TIMER_7
+
+#define CYC_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC \
+		      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+#define MAX_TICKS ((0x00ffffffu - CYC_PER_TICK) / CYC_PER_TICK)
+
+enum _EXT_TIMER_PRESCALE_TYPE_ {
+	ET_PSR_32K,
+	ET_PSR_1K,
+	ET_PSR_32,
+	ET_PSR_8M,
+};
+
+enum _EXT_TIMER_IDX_ {
+	EXT_TIMER_3 = 0,        /* ctimer */
+	EXT_TIMER_4,            /* ctimer */
+	EXT_TIMER_5,            /* etimer */
+	EXT_TIMER_6,            /* NULL */
+	EXT_TIMER_7,            /* rtimer */
+	EXT_TIMER_8             /* NULL */
+};
+
+/* be careful of overflow issue */
+#define MILLI_SEC_TO_COUNT(hz, ms) ((hz) * (ms) / 1000)
+#define MICRO_SEC_TO_COUNT(hz, us) ((hz) * (us) / 1000000)
+
+#define TIMER_STOP(Idx)          (ECReg(EXT_CTL_B + ((Idx) * 8)) &= ~BIT0)
+#define TIMER_START(Idx)         (ECReg(EXT_CTL_B + ((Idx) * 8)) |= BIT0)
+#define TIMER_RESTART(Idx)	  \
+	do {			  \
+		TIMER_STOP(Idx);  \
+		TIMER_START(Idx); \
+	} while (0)
+#define TIMER_RELOAD(Idx, Cnt)						  \
+	do {								  \
+		TIMER_START((Idx));					  \
+		ECReg(EXT_LH3R_B + ((Idx) * 8)) = (((Cnt) >> 24) & 0xFF); \
+		ECReg(EXT_LH2R_B + ((Idx) * 8)) = (((Cnt) >> 16) & 0xFF); \
+		ECReg(EXT_LHR_B  + ((Idx) * 8)) = (((Cnt) >>  8) & 0xFF); \
+		ECReg(EXT_LLR_B  + ((Idx) * 8)) = (((Cnt) >>  0) & 0xFF); \
+	} while (0)
+/* The macro can be used only for Timer #3 ~ #7 */
+#define TIMER_DISABLE(Idx)      CLEAR_MASK_EXT(IER19, (BIT(3 + (Idx))))
+#define TIMER_ENABLE(Idx)               SET_MASK_EXT(IER19, (BIT(3 + (Idx))))
+#define TIMER_CLEAR_FLAG(Idx)       (ISR19 = BIT(3 + (Idx)))
+#define TIMER_CHECK_FLAG(Idx)       (ISR19 & BIT(3 + (Idx)))
+#define TIMER_WAIT(Idx)         (!(ISR19 & BIT(3 + (Idx))))
```

Comment:
most of these macros can be static functions

trust the compiler does the right thing ;)